server/routes/termdb.cluster.ts

```
@@ -267,95 +268,311 @@ export async function validate_query_geneExpression(ds: any, genome: any) {
-		if (limitSamples?.size == 0) {
-			// got 0 sample after filtering, must still return expected structure with no data
-			return { term2sample2value: new Map(), byTermId: {}, bySampleId: {} }
+		// Debug output to understand what we're getting back
+		// console.log('Result structure:', JSON.stringify(result, null, 2).substring(0, 5000) + '...');
+
+		// Check if the result exists and contains sample data
+		if (!result || Object.keys(result).length === 0) {
+			throw new Error('Failed to retrieve expression data: Empty or missing response')
 		}
 
-		// has at least 1 sample passing filter and with exp data
-		// TODO what if there's just 1 sample not enough for clustering?
-		const bySampleId = {}
-		const samples = q.samples || []
-		if (limitSamples) {
-			for (const sid of limitSamples) {
-				bySampleId[sid] = { label: ds.cohort.termdb.q.id2sampleName(sid) }
-			}
-		} else {
-			// use all samples with exp data
-			for (const sid of samples) {
-				bySampleId[sid] = { label: ds.cohort.termdb.q.id2sampleName(sid) }
+		return result
+	} catch (error) {
+		console.error(`Error querying gene expression for ${geneName}`)
+		throw error
+	}
+}
+
+/**
+ * Validate and prepare a gene expression query
+ * This function handles both HDF5 and tabix file formats
+ *
+ * @param q - The gene expression query
+ * @param ds - Dataset information
+ * @param genome - Genome information
+ */
+async function validateNative(q: GeneExpressionQueryNative, ds: any, genome: any) {
+	// Determine whether we're handling an HDF5 file or a tabix file
+	if (q.hdf5File === true) {
+		// HDF5 file handling branch
+		const h5FilePath = q.file
+
+		// Join with master directory if needed
+		if (!h5FilePath.startsWith(serverconfig.tpmasterdir)) {
+			q.file = path.join(serverconfig.tpmasterdir, h5FilePath)
+		}
+
+		if (!q.samples) q.samples = []
+
+		// Validate that the HDF5 file exists
+		await utils.file_is_readable(q.file).catch(_error => {
```

simply `await file_is()` and no need to catch, just let it throw, and it will be caught upstream

---

Will remove the catch. When not using `utils `I get an error from typescript about not being able to find the name `file_is_readable.ts`. The various functions got imported as `utils`. If you would prefer not to have `utils.file_is_readable` here I can explicitly import just `file_is_readable`

---

Addressed in [18c6b8e](https://github.com/stjude/proteinpaint/pull/2998/commits/18c6b8e6ecfcfd851f6c14e39aac9c6a9078366f)